tensorflow_probability/python/distributions/distribution.py

```diff
@@ -909,12 +909,34 @@ def _call_sample_n(self, sample_shape, seed, name, **kwargs):
       if JAX_MODE and seed is None:
         raise ValueError('Must provide JAX PRNGKey as `dist.sample(seed=.)`')
       sample_shape = tf.cast(sample_shape, tf.int32, name='sample_shape')
+
+      if "conditional" in kwargs.keys():
```

Comment:
Is this doing a kind of cross-product between conditioning and number of samples? That's very unlike the rest of TFP, where we'd broadcast these things. Can you explain what the goal of this is? I think there's a few different semantics that are possible here, and I'd like to know which one you had in mind.

Reply:
Yes, say we want 10 samples from the P(x|d_i), but we have 100 different sets of conditional parameters (d_i | i \elem [1, 100]) then we would want to have 1000 samples returned, but shaped as [10, 100]. I think there are lots of interesting options to consider here too. For example we might want to have the same random samples of x, but under different conditions, so that would be 10 random numbers in total which would all be conditioned to allow us to do comparison. The other option is that you might have 100 different conditions and want 10 random samples from each and so 1000 different samples from x would be needed. I actually added this whole part because of the shape checking on the output failing with different numbers of conditional inputs which is why I couldn't just use broadcasting. I'm happy to (would rather) change this to make it a less hacky solution. But it was the best I could come up with whilst needing to get pydelfi working.

Reply:
I've given this a bit of thought. So the primary issue is that this shape issue is not specific to conditioning. Any bijector, in principle, can add additional batch dimensions, and these added dimensions behave exactly like the batch dimensions on the conditioning. E.g. consider this:

```python
base = tfd.Normal(0., 1.)
scale = tfb.Shift([10., 100.])

transformed = scale(base)
x = transformed.sample(1)
# Raises an error:
# Tensor's shape (2,) is not compatible with supplied shape (1,)
```

I assume that's the error you hit. To get around it, we can match the number of samples, to the conditioned batch dimensions:

```python
x3 = transformed.sample(2)
# tf.Tensor([  8.294859 100.95336 ], shape=(2,), dtype=float32)

x4 = transformed.sample([4, 2])
# tf.Tensor(
# [[ 10.168862  99.26204 ]
#  [ 10.265037 100.74291 ]
#  [ 11.301426  98.726166]
#  [ 10.991593  99.583084]], shape=(4, 2), dtype=float32)
```
Note that the randomness is not shared between the elements in the samples.

We can effect sharing of randomness if we add a 'repeat' bijector, something like this:

```python
def repeat(x):
  return tf.repeat(x[..., tf.newaxis], 2, axis=-1)

def fwd_event_shape(s):
  return s.concatenate([2])

rep_b = tfb.Inline(repeat, forward_event_shape_fn=fwd_event_shape, forward_min_event_ndims=0, inverse_min_event_ndims=1)
rep_base = rep_b(base)
rep_transformed = scale(rep_base)

x5 = rep_transformed.sample(1)
# tf.Tensor([[ 10.948769 100.94877 ]], shape=(1, 2), dtype=float32)
```

Now, I'm describing the way things are, without saying that it's good or bad. My question to you is, ignoring some minor inconvenience, is your use case still supported if you adjust your `sample` calls to specify the broadcastable number of samples? Do you actually want to be able to share randomness (in which case we can add a real `repeat` bijector).

If not, we probably can come up with a solution, but I think it'll need to handle the batch-adding bijectors above, as they're fundamentally the same as these conditional bijectors you're adding.

Reply:
I think for now I am happy to just try the broadcasting sampling, although the repeated sampling would be nice for certain problems I'm sure. At the moment, if I just try using this broadcasting I get an error like
`WARNING:tensorflow:Model was constructed with shape (None, 6) for input Tensor("input_1:0", shape=(None, 6), dtype=float32), but it was called on an input with incompatible shape (12, 6, 6).`
for 
`
distribution.conditional_sample([12, 10], conditional=tf.convert_to_tensor(np.random.normal(0, 1, (10, 6))))
`
where
`
distribution.conditional_sample = lambda a, conditional=None : distribution.sample(a, bijector_kwargs=put_conditional(conditional))
`
and
`
put_conditional = lambda conditional : dict(
        zip(["MADE_{}".format(i) for i in range(n_mades)], 
            [{"conditional": conditional} for i in range(n_mades)]))
`

Reply:
I think that warning is probably harmless, (and it'd be a pain to fix on our end, maybe we could ask Keras folks to be more tolerant of added batch dimensions). A more important issue is that the output reshape doesn't do the right thing, you need this sort of change:

```diff
--- a/tensorflow_probability/python/bijectors/masked_autoregressive.py
+++ b/tensorflow_probability/python/bijectors/masked_autoregressive.py
@@ -1027,11 +1027,17 @@ class AutoregressiveNetwork(tf.keras.layers.Layer):
           raise ValueError('`conditional` must be passed as a named arguement')
         conditional = tf.convert_to_tensor(conditional, dtype=self.dtype,
                                            name="conditional")
+        conditional_batch_shape = tf.shape(conditional)[:-1]
         if tensorshape_util.rank(conditional.shape) == 1:
           conditional = conditional[tf.newaxis, ...]
         x = [x, conditional]
+        output_shape = tf.concat(
+            [tf.broadcast_dynamic_shape(conditional_batch_shape, input_shape[:-1]),
+              input_shape[-1:]], axis=0)
+      else:
+        output_shape = input_shape
       return tf.reshape(self._network(x),
-                        tf.concat([input_shape, [self._params]], axis=0))
+                        tf.concat([output_shape, [self._params]], axis=0))
```

Then, code like this appears to run without error (I haven't verified that the outputs are correct though):

```python
import itertools

import tensorflow as tf
import tensorflow_probability as tfp

tfb = tfp.bijectors

made = tfb.AutoregressiveNetwork(2, [3], conditional=True, conditional_shape=[4])

input_shapes = [
  [3],
  [1, 3],
  [2, 3],
  [1, 2, 3],
  [2, 1, 3],
  [2, 2, 3],
]
cond_shapes = [
  [4],
  [1, 4],
  [2, 4],
  [1, 2, 4],
  [2, 1, 4],
  [2, 2, 4], 
]
    
for input_shape, cond_shape in itertools.product(input_shapes, cond_shapes):
  print(input_shape, cond_shape)
  print(made(tf.ones(input_shape), conditional=tf.ones(cond_shape)).shape)
```

We'll want a unittest like that, to be sure.

Reply:
Great, I actually tracked down a bug which caused the bad broadcasting but the error remained, so I like this way much more. It definitely seems to be doing the correct thing now.